### DISCUSSION TITLE
LLM inference server performances comparison llama.cpp / TGI / vLLM

### DISCUSSION THREAD

**Discussion by phymbert** [ggml-org/llama.cpp#6730] [General]
## Performances and improvment area

This thread objective is to gather `llama.cpp` performance 📈 and improvement ideas💡against other popular LLM inference
frameworks, especially on the `CUDA` backend. Let's try to fill the gap 🚀

1. [Hugging Face TGI](https://github.com/huggingface/text-generation-inference): A Rust, Python and gRPC server for text
   generation inference.
2. [vLLM](https://github.com/vllm-project/vllm): Easy, fast, and cheap LLM serving for everyone.

I have run a couple of benchmarks from the OpenAI `/chat/completions` endpoint client point of view
using [JMeter](https://jmeter.apache.org/) on 2 A100 with `mixtral8x7b` and a fine tune `llama70b` models.

Note 1: from the client point of view, it is not possible to get accurate PP and TG because, first you need steaming
enabled and then PP will always include one generated token. So easier to compare the total tokens of the transactions
in `completions.usage`.

Note 2: from a performance tests server point of view, we generally consider following metrics:

- `iterations`: total request successfully completed during the test
- `prompt tokens`: average prompt tokens per request, same by iteration number for all tests
- `generated tokens`: average generated tokens per request
- `RPM`: Requests Per Minute
- `latency`: Duration of the http request in seconds
- `PP+TG`: total tokens http clients send and receive per second
- `errors`: number of request in errors during the test from the client point of view, it can be http timeout,
  connection close. It is not necessarily caused by the server.

#### Context size

The transaction tokens context here is:

|                  | min | avg | max |
|------------------|-----|-----|-----|
| Prompt tokens    | 12  | 155 | 512 |
| Generated tokens | 38  | 133 | 996 |

#### llama70b @ eedd42e

| metric           | users | duration | llama.cpp | vLLM   | TGI    |
|------------------|-------|----------|-----------|--------|--------|
| iterations       | 32    | 30m      | 1514      | 4 734  | 4 448  |
| prompt tokens    | 32    | 30m      | 155.19    | 125.58 | 138.66 |
| generated tokens | 32    | 30m      | 132.71    | 144.61 | 125.66 |
| RPM              | 32    | 30m      | 40.92     | 147.94 | 139.00 |
| latency          | 32    | 30m      | 65.10     | 17.06  | 16.79  |
| PP+TG/s          | 32    | 30m      | 196.34    | 649.46 | 612.33 |
| errors           | 32    | 30m      | 0.19%     | 0.09%  | 0.05%  |
| iterations       | 1     | 10m      | 48        | 99     | 85     |
| prompt tokens    | 1     | 10m      | 120.31    | 116.07 | 135.53 |
| generated tokens | 1     | 10m      | 115.56    | 121.28 | 124.69 |
| RPM              | 1     | 10m      | 4.36      | 9.00   | 7.73   |
| latency          | 1     | 10m      | 12.83     | 6.41   | 7.30   |
| PP+TG            | 1     | 10m      | 17.15     | 35.60  | 33.51  |
| errors           | 1     | 10m      | 0%        | 0%     | 0%     |

##### llama.cpp configuration

```shell
server  --model myllama70b-f16-00001-of-00010.gguf \
        --ctx-size 32768 \
        --n-predict 4096 \
        --n-gpu-layers 81 \
        --batch-size 4096 \
        --ubatch-size 256 \
        --parallel 1|32 \
        --metrics \
        --log-format text
```

##### vLLM configuration

```shell
python -m vllm.entrypoint.openai.api_server \
   --model /models/myllama70b \
   --tensor-parallel-size 2
```

##### TGI Configuration

Please note how it is easy:

```shell
text-generation-launch --model-id /models/myllama70b
```

#### mixtral8x7b

| metric           | users | duration | llama.cpp | vLLM     | TGI      |
|------------------|-------|----------|-----------|----------|----------|
| iterations       | 32    | 30m      | 4 152     | 10 541   | 10 849   |
| prompt tokens    | 32    | 30m      | 83.56     | 97.63    | 98.63    |
| generated tokens | 32    | 30m      | 110.27    | 166.31   | 98.18    |
| RPM              | 32    | 30m      | 129.75    | 329.41   | 339.03   |
| latency          | 32    | 30m      | 20.90     | 6.04     | 5.83     |
| PP+TG/s          | 32    | 30m      | 409.81    | 1 449.13 | 1 487.45 |
| errors           | 32    | 30m      | 1.29%     | 0.06%    | 0.05%    |
| iterations       | 1     | 10m      | 219       | 439      | 430      |
| prompt tokens    | 1     | 10m      | 79.24     | 94.09    | 98.64    |
| generated tokens | 1     | 10m      | 103.02    | 107.65   | 108.14   |
| RPM              | 1     | 10m      | 19.91     | 39.91    | 39.09    |
| PP+TG            | 1     | 10m      | 60.48     | 134.19   | 134.72   |
| errors           | 1     | 10m      | 0%        | 0%       | 0%       |

##### llama.cpp configuration @ 137fbb8

```shell
server  --model mixtral-8x7b-instruct-f16-00001-of-00010.gguf \
        --ctx-size 131072 \
        --n-predict 4096 \
        --n-gpu-layers 33 \
        --batch-size 4096 \
        --ubatch-size 256 \
        --parallel 1|32 \
        --metrics \
        --log-format text
```

Magically vLLM and TGI configuration are not changed.

### Area of improvements of `llama.cpp`

Please @ggerganov edit as will

- #6505
- #5021
- #6502
- Automatic KV Cache size detection
- Automatic batch size based on the underlying hardware
- #6607

**Comment by phymbert** [comment]
@JohannesGaessler @slaren as you are the main contributors on CUDA backend feel free to highlight or amend any hypothesis. Thanks a lot for your impressive job here.

**Reply by JohannesGaessler** [reply]
It's 80% a matter of what areas the optimization effort has been going towards. I really don't care about the performance of FP16 or many users using a single server. What I care about is reducing the cost of a single user running the best available models at acceptable speeds.

**Reply by phymbert** [reply]
It's clearly in the manifesto of `llama.cpp` to focus on inference at edge, no problem with that. That's why I have added the 1 user scenario. Regarding F16, first I have a bad feeling with quantum models but this is subjective straight F16 always answers better for me, coding assistant, translation or general questions.

Secondly, vLLM and TGI have now good support of quantized model, I can add another round on Q4 but the gap will be definitely the same.

The main question is why our CUDA implementation is twice slower on all benchmark ?

**Reply by JohannesGaessler** [reply]
>Secondly, vLLM and TGI has now good support of quantized model, I can add another round on Q4 but the gap will be definitely the same.

Q4 != Q4. What matters is not some arbitrary quant number but the tradeoff between quality, speed, and VRAM use. I could write a very fast "Q4" or "Q8" format that would not be worth using because the quality would be terrible relative to the VRAM use. 

>The main question is why our CUDA implementation is twice slower on all benchmark ?

The biggest issue with llama.cpp FP16 right now is probably that there is no support for FP16 compute buffers. It's therefore necessary to do FP16 <-> FP32 conversions and to do some calculation using the slower FP32 even if the precision and range are not needed.

Also llama.cpp would need support for fusing tensors whenever possible.

**Reply by slaren** [reply]
With `llama-bench`, I get 55 t/s with llama 7B F16. Considering the model size alone, that requires about 700 GB/s of memory bandwidth, about 70% of the theoretical bandwidth of this 3090 Ti. Getting double the performance, at least on generation, does not seem possible on this hardware. The results may be worse on an A100, and that wouldn't be unexpected since nobody has spent any time optimizing for that specific hardware.

**Reply by ggerganov** [reply]
@phymbert Can you run `llama-bench` with the 70B model and show the numbers? Also, `llama.cpp` runs with pipeline parallelism by default, while the other frameworks likely use tensor parallelism. I think `llama.cpp` TP is not as good atm, but you can give it a try `--split-mode row`

**Reply by JohannesGaessler** [reply]
I forgot: a large part of the issue probably has to do with multi GPU optimization. Since the A100s are presumably connected via NVLink, try compiling with `LLAMA_CUDA_PEER_MAX_BATCH_SIZE=9999999`. It's unfortunately not possible to query NVLink status from within CUDA so peer access is by default not always used. Also with NVLink `--split-mode row` should be much faster.

**Reply by phymbert** [reply]
Isee, I had also benchmarked PHI2 Q4 on my poor RTX 3050 here #5941 and vLLM has a non negligeable gap.

**Reply by slaren** [reply]
Peer access will never be enabled when using the default `-sm layer`, it is only done when on split buffer. Still, the amount data that's transferred between GPUs is very small, so I wouldn't expect any meaningful improvement even with NVLink.

**Reply by phymbert** [reply]
@JohannesGaessler I am still on PCI here, but H100 NVLink will arrive for sure. Can you add `--flash-attn` in `llama-bench` in your branch please ?

**Reply by JohannesGaessler** [reply]
Just change the default value?

**Reply by phymbert** [reply]
#### llama-bench on 2 A100 80GB PCIe compute capabilities 8.0, VMM: yes

@ggerganov here it is @ 44ca5764d621d8693f8f8c01b9b920d7620c9076:

| model         | size       | params  | backend | ngl | n_batch | n_ubatch | sm    | test   | t/s     |
|---------------|------------|---------|---------|-----|---------|----------|-------|--------|---------|
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | layer | pp 512 | 1280.50 |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | layer | tg 128 | 7.67    |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | layer | pp 512 | 1147.39 |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | layer | tg 128 | 7.67    |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | row   | pp 512 | 242.49  |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | row   | tg 128 | 9.97    |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | row   | pp 512 | 293.74  |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | row   | tg 128 | 9.98    |

@ 44ca5764d621d8693f8f8c01b9b920d7620c9076 w/ flash_attn = true;
EDIT: updated on 2024/04/18 with flash_attn = true, in llama.cpp...

| model         | size       | params  | backend | ngl | n_batch | n_ubatch | sm    | test   | t/s     |
|---------------|------------|---------|---------|-----|---------|----------|-------|--------|---------|
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | layer | pp 512 | 1404.14 |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | layer | tg 128 | 7.82    |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | layer | pp 512 | 1228.39 |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | layer | tg 128 | 7.82    |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | row   | pp 512 | 252.75  |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 256      | row   | tg 128 | 10.25   |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | row   | pp 512 | 299.33  |
| llama 70B F16 | 128.48 GiB | 68.98 B | CUDA    | 81  | 4096    | 512      | row   | tg 128 | 10.26   |

**Reply by phymbert** [reply]
@ggerganov Is it me or it does not flash ?

**Reply by ggerganov** [reply]
Did you change `flash_attn = true` in `common.h`? You have to change the one in `llama.cpp`

**Reply by phymbert** [reply]
Thanks, figures updated.
What are the action plan if any ? Should we try to improve the split mode row ?
I will test the same on H100 asap. Focusing on A100 is not important as it is out of sale anyway.

**Reply by ggerganov** [reply]
The `PP+TG` metric is heavily biased by the length of the generated responses. If for some reason `llama.cpp` tends to generate much longer responses than the others, it would affect negatively the metric because TG speed is orders of magnitude lower than PP speed.

You can mitigate this to some extend if you limit the max generated response some small number (for example max 32 tokens generated per request)

**Reply by phymbert** [reply]
I audited the responses, and I also have the input and output tokens generated. They are equivalent, but fair enough I will update the table with the avg prompts and tokens size.

**Reply by phymbert** [reply]
@ggerganov updated, I think this is not the cause, and TGI and vLLM look quite close in term of performances, with the same benchmark framework for all of 3.

**Reply by ggerganov** [reply]
Yup, looks like the efficiency is no there yet. As mentioned earlier, A100 support was never a priority and tensor parallelism potentially needs work. Though from pure bandwidth calculations, we are at about 50% utilization (assuming 2TB/s MBW for A100), so probably the former is the dominating factor. Still, quite useful information - thanks for taking the time

I'm interested in one more benchmark if you have the capacity: `Q8_0` 70B and 7x8B on a single A100 with all 3 frameworks

**Reply by phymbert** [reply]
I would need to have a mapping between our quantization methods and the python ones:

vLLM supports: [GPTQ](https://arxiv.org/abs/2210.17323), [AWQ](https://arxiv.org/abs/2306.00978), [SqueezeLLM](https://arxiv.org/abs/2306.07629), FP8 KV Cache
TGI supports: [bitsandbytes](https://github.com/TimDettmers/bitsandbytes) [GPT-Q](https://arxiv.org/abs/2210.17323) [EETQ](https://github.com/NetEase-FuXi/EETQ) [AWQ](https://github.com/casper-hansen/AutoAWQ)

If you guide me a little on how I can convert a bfloat16 HF model to the quantized version, I can give it a try on my spare time.

**Reply by ouvaa** [reply]
@phymbert for TGI and vLLM, what file format was used? have you tested which file format is faster?

basically was wondering what is the fastest way to do inference with what file format

**Reply by phymbert** [reply]
What are you calling file format, please ? I just took the files on HF, safetensors bfloat16

**Reply by phymbert** [reply]
> basically was wondering what is the fastest way to do inference with what file format

llama.cpp on GGUF of course

**Reply by ggerganov** [reply]
> If you guide me a little on how I can convert a bfloat16 HF model to the quantized version, I can give it a try on my spare time.

I'm not familiar with the other frameworks, so would be of little help here. No worries, not very important

**Reply by OB-SPrince** [reply]
I think the llama.cpp need to use the real model, and not a quantized gguf version. Then it would be proper comparison.

**Reply by phymbert** [reply]
Please develop

**Comment by aneeshmb02** [comment]
I have the same objective. What data/prompt did you use and can I use the same with [llama-bench](https://github.com/ggerganov/llama.cpp/tree/master/examples/llama-bench)? 
Started the same discussion at #7195.

**Comment by uicomponent** [comment]
@phymbert Can you please add updated test results? Based on details in linked threads inference server performance was significantly improved. It would be nice to compare results before and after here.

**Reply by phymbert** [reply]
It supposed to have a continuuous performance github action but it has been disabled. Feel free to fix it

**Comment by hitdra** [comment]
Hello, I'm new to this area and recently wanted to measure llamacpp throughput on an a100 video card. I would like to ask you, what about sending a request to llamacpp server after opening its service and then measuring the throughput of llamacpp?

If so, can you refer me to your code for measuring these metrics?

**Reply by VJHack** [reply]
Start `llama-server` with `--metrics` then [hit `/metrics` endpoint.](https://github.com/ggerganov/llama.cpp/tree/master/examples/server#:~:text=GET%20/metrics%3A%20Prometheus%20compatible%20metrics%20exporter)

**Reply by hitdra** [reply]
Thank you very much for your reply, I still don't quite get it, could you please give me a specific example if you can?

**Reply by hitdra** [reply]
> Start `llama-server` with `--metrics` then [hit `/metrics` endpoint.](https://github.com/ggerganov/llama.cpp/tree/master/examples/server#:~:text=GET%20/metrics%3A%20Prometheus%20compatible%20metrics%20exporter)

Sorry, I have another query for you. How do I send so many requests in the dataset to the llamacpp server? With a one by one curl command?

**Comment by OB-SPrince** [comment]
Use an AI to develop a load testing script, by showing an example of how you use curl.

I have a repo for this https://GitHub.com/SolidRusT/srt-inference-perf

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by raulod** [comment]
would it be possible for you to create benchmark for CPU only since vllm supports CPU now

**Reply by phymbert** [reply]
You can use the bench tool in the server example against any openai API compatible solution. Feel free to share your findings.

**Comment by foldl** [comment]
FYI: 

[LLM-Inference-Bench: Inference Benchmarking of Large Language Models on AI Accelerators](https://arxiv.org/html/2411.00136v1)

This paper includes some benchmarks of `llama.cpp`.

**Comment by lingster** [comment]
Interesting thread has anyone run any profiler perf tools such as val grind or gprof to see where bottlenecks are within llamacpp?

**Comment by AGenchev** [comment]
In my setup llama.cpp performs **faster** than vLLM, model is thinking model Qwen QwQ32B, GPU#2 is A100/80GB, single request:
Vllm start:
```bash
export CUDA_VISIBLE_DEVICES=2 
vllm serve weights/QwQ_Q8_0.gguf  \
  --tensor-parallel-size 1  \
  --pipeline-parallel-size 1  \
  --tokenizer weights/QwQ-32B \
  --port 1027
```
llama.cpp start:
```bash
export CUDA_VISIBLE_DEVICES=2 
GGML_CUDA_ENABLE_UNIFIED_MEMORY=1 taskset -c 0-47 ~/src/llama.cpp/build/bin/llama-server \
  -m '/home/gele/LLMs/weights/QwQ_Q8_0.gguf'  \
  --temp 0.3 --top-k 40 --top-p 0.9 --repeat-penalty 1.1 -n 8192 --ctx-size 40960 \
  --host 192.168.30.5 --port 1027 --cpu-range 0-47 --cpu-strict 1 -ngl 70
```
timing of second run (first run to warm-up,same task):
VLLM: 2m7.356s
llama.cpp: 1m42.747s
But it seems vLLM gives more times the better result.

**Reply by suparious** [reply]
Ok, now try again, using parallel requests, like maybe even just 10-20 requests at the same time. That is what vLLM is meant for, not for single user inference.

**Reply by phymbert** [reply]
Try changing vllm `--tensor-parallel-size` according to visible devices... but indeed, PP is great with llama.cpp while Ampere and Hooper nvidia arch are not targeted for.

**Reply by suparious** [reply]
What I was trying to highlight, was how vLLM uses batched inference, where it takes multiple requests, batches them together, then runs a single inference. I was not referring to tensor parallelization, however, to your point, this would dramatically increase the performance of using parallel requests.

**Reply by AGenchev** [reply]
So, how we measure/test is very important. 
You didn't clearly state that you have batched the requests to squeeze maximum achievable parallelization so I assumed the "iterations" were executed sequentially within the given time as the definition of iteration implies this. 
That's why I made a simple, sequential test which made your results to look wrong.

To make parallel requests, I need to modify my client code to use non-blocking sockets and complex error handling (retry on error & combine with next data, order results & logging), because my case requires the results to be ordered. Without reworking code, I get more iterations/time with llama.cpp.

**Comment by adhusch** [comment]
Very interesting. Thanks for sharing.

Any plans to redo the experiment given all the improvements made over the last months?
